### PR TITLE
Migrate key value pairs from Pegasus to DynamoDB [#101145644]

### PR DESCRIPTION
**Pull request by davidsbailey**
For now, keep writing key value pairs (a.k.a. properties) to Pegasus, too.

Addresses: https://www.pivotaltracker.com/story/show/101145644

Migration plan: https://docs.google.com/document/d/1I9qFlhqdscOtqmrjq8Qz9gAMlHW01keHWZ482HirXUo/edit#

Done in this PR:
- write a migration to populate `CDO.dynamo_properties_table` in DynamoDB from `app_properties` in Pegasus
- change the behavior of `CDO.use_dynamo_properties` to write to both places and read from DynamoDB.
- set `CDO.use_dynamo_properties` to be true on staging/test/prod

Manual test steps performed locally (from the above doc):
- clear `use_dynamo_properties` in locals.yml 
- write two test apps which store properties data (in Pegasus)
- set `use_dynamo_properties` in locals.yml 
- run the migration locally
- manually verify that all properties data has been copied from Pegasus to DynamoDB
- load the test apps and verify that (1) they have all existing data in both places and (2) new data is written to both places
- rollback migration, roll forward migration, verify no more changes to DynamoDB
- `shared$ ruby -Itest test/test_properties.rb`

### CODE REVIEW

**Comment by davidsbailey** [comment]
It would be great to have a few sets of eyes on this since doing it wrong could mess up user data:
- @Bjvanminnen for middleware code changes and general review
- @philbogle for migration plan
- @laurelfan for migration code sanity check

Also FYI @joshlory and @sfilman .

**Comment by Bjvanminnen** [comment]
changes lgtm

File: pegasus/migrations/79_migrate_properties_to_dynamodb.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+require 'cdo/hip_chat'
+require 'properties_api'
+
+Sequel.migration do
+  up do
+    HipChat.log 'Copying <b>>key value pairs</b> to DynamoDB...'
```
Comment (laurelfan) [inline pegasus/migrations/79_migrate_properties_to_dynamodb.rb]: I think you have an extra &gt; here

File: pegasus/migrations/79_migrate_properties_to_dynamodb.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+    # There are currently < 1000 rows in production. Don't bother batching.
+    rows = from(:app_properties)
+    rows.each do |row|
+      value = PropertyBag.parse_value(row[:value])
+      property_bag = DynamoPropertyBag.new(row[:app_id], row[:storage_id])
```
Comment (laurelfan) [inline pegasus/migrations/79_migrate_properties_to_dynamodb.rb:18]: This is a create or update sort of thing?

File: pegasus/migrations/79_migrate_properties_to_dynamodb.rb [hunk 3]
```diff
@@ -0,0 +1,22 @@
+require 'cdo/hip_chat'
```
Comment (laurelfan) [inline pegasus/migrations/79_migrate_properties_to_dynamodb.rb:1]: migration LGTM

File: pegasus/migrations/79_migrate_properties_to_dynamodb.rb (hunk 2 above)
Comment (davidsbailey) [inline pegasus/migrations/79_migrate_properties_to_dynamodb.rb:18]: Yes. `DynamoPropertyBag.set` ends up being a "put" to the DynamoDB API: https://github.com/code-dot-org/code-dot-org/blob/staging/shared/middleware/helpers/property_bag.rb#L106

File: pegasus/migrations/79_migrate_properties_to_dynamodb.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+require 'properties_api'
+
+Sequel.migration do
+  up do
+    HipChat.log 'Copying <b>key value pairs</b> to DynamoDB...'
+
+    # There are currently < 1000 rows in production. Don't bother batching.
+    rows = from(:app_properties)
+    rows.each do |row|
+      value = PropertyBag.parse_value(row[:value])
```
Comment (philbogle) [inline pegasus/migrations/79_migrate_properties_to_dynamodb.rb]: Can parse_value fail?

**Comment by philbogle** [comment]
LGTM

File: pegasus/migrations/79_migrate_properties_to_dynamodb.rb (hunk 4 above)
Comment (davidsbailey) [inline pegasus/migrations/79_migrate_properties_to_dynamodb.rb]: Good catch Phil. It turns out there were 3 rows with bad JSON, at least one of which was due to  being truncated:

```
ubuntu@production-daemon:~/production$ bin/dashboard-console 
irb(main):035:0>     PEGASUS_DB[:app_properties].each do |row|
irb(main):036:1*       begin
irb(main):037:2*         value = PropertyBag.parse_value(row[:value])
irb(main):038:2>       rescue JSON::ParserError
irb(main):039:2>         puts "<b>Skipping bad key value pair with id #{row[:id]}</b>."
irb(main):040:2>         next
irb(main):041:2>       end
irb(main):042:1>     end
<b>Skipping bad key value pair with id 219</b>.
<b>Skipping bad key value pair with id 295</b>.
<b>Skipping bad key value pair with id 327</b>.
=> #<Sequel::Mysql2::Dataset: "SELECT * FROM `app_properties`">
```

I've updated the code to log this error and continue.

**Comment by davidsbailey** [comment]
This migration has completed successfully. I have a follow-up item to gracefully handle values which are too large: https://www.pivotaltracker.com/story/show/108811134

**Comment by davidsbailey** [comment]
FYI, no HipChat logging messages were written during the migration, however I did manually verify the new row count in `production_tables` in DynamoDB.

**Pull request merged**